src/compile.rs

```diff
@@ -1528,57 +1528,62 @@ fn generate_type_no_parent_errors(missed_types: Vec<&TypeInfo>, types: &TypeMap)
     ret
 }
 
-fn get_synthetic_resource_name(
+pub fn get_synthetic_resource_name(
     dom_name: &CascadeString,
     associated_resource: &CascadeString,
 ) -> CascadeString {
-    format!("{}.{}", dom_name, associated_resource).into()
+    let cs_name = format!("{}.{}", dom_name, associated_resource);
+    // We keep the range of the *resource* part specifically, which should always be where this
+    // resource was defined
+    match associated_resource.get_range() {
+        Some(range) => CascadeString::new(cs_name, range),
+        None => CascadeString::from(cs_name),
+    }
 }
 
 fn create_synthetic_resource(
     types: &TypeMap,
     dom_info: &TypeInfo,
     associated_parents: &Vec<&TypeInfo>,
-    class: &TypeInfo,
+    classes: Vec<&TypeInfo>,
     class_string: &CascadeString,
     global_exprs: &mut HashSet<Expression>,
     extend_annotations: &BTreeMap<CascadeString, BTreeSet<AnnotationInfo>>,
 ) -> Result<CascadeString, ErrorItem> {
-    if !class.is_resource(types) {
+    if !classes.iter().all(|c| c.is_resource(types)) {
+        // The key thing here is the ranges of each class, which the associated resource struct
+        // owns.  Perhaps just passing it in wholesale makes sense?
         return Err(ErrorItem::make_compile_or_internal_error(
             "not a resource",
             dom_info.declaration_file.as_ref(),
-            class_string.get_range(),
+            classes
+                .iter()
+                .find(|c| !c.is_resource(types))
+                .and_then(|c| c.name.get_range()),
```

Comment:
I _think_ this will just return the first class that is a not a resource in the Vec.  I think you should be able to use maybe_add_additional_message to get all of them in one pass.